sdk/daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/crypto/Hash.scala

```diff
@@ -180,18 +212,248 @@ object Hash {
+      // If the node is version agnostic, the builder doesn't matter so pick V1
+      // This only applies to rollback nodes
+      case None => new NodeBuilderV1(messageDigest, valueBuilder)
+    }
   }
 
-  private final class HashMacBuilder(key: Hash) extends Builder {
+  private final class NodeBuilderV1(
+      messageDigest: MessageDigestWithContext,
+      override protected val valueBuilder: ValueHashBuilder,
+  ) extends NodeBuilder(NodeHashVersion.V1, messageDigest, valueBuilder) {
+    private def addCreateNode(create: Node.Create): this.type = {
+      addContext("Create Node")
+        .withContext("Contract Id")(_.addWithValueBuilder(create.coid, _.addCid))
+        .withContext("Package Name")(_.addString(create.packageName))
+        .withContext("Template Id")(_.addIdentifier(create.templateId))
+        .withContext("Arg")(_.addWithValueBuilder(create.arg, _.addTypedValue))
+        .withContext("Signatories")(_.addStringSet(create.signatories))
+        .withContext("Stakeholders")(_.addStringSet(create.stakeholders))
+    }
+
+    private def addFetchNode(fetch: Node.Fetch): NodeBuilder = {
+      addContext("Fetch Node")
+        .withContext("Contract Id")(_.addWithValueBuilder(fetch.coid, _.addCid))
+        .withContext("Package Name")(_.addString(fetch.packageName))
+        .withContext("Template Id")(_.addIdentifier(fetch.templateId))
+        .withContext("Signatories")(_.addStringSet(fetch.signatories))
+        .withContext("Stakeholders")(_.addStringSet(fetch.stakeholders))
+        .withContext("Acting Parties")(_.addStringSet(fetch.actingParties))
+        .withContext("Interface Id")(_.addOptional(fetch.interfaceId, _.addIdentifier))
+    }
+
+    private def addExerciseNode(exercise: Node.Exercise, nodes: Map[NodeId, Node]): NodeBuilder = {
+      addContext("Exercise Node")
+        .withContext("Contract Id")(_.addWithValueBuilder(exercise.targetCoid, _.addCid))
+        .withContext("Package Name")(_.addString(exercise.packageName))
+        .withContext("Template Id")(_.addIdentifier(exercise.templateId))
+        .withContext("Signatories")(_.addStringSet(exercise.signatories))
+        .withContext("Stakeholders")(_.addStringSet(exercise.stakeholders))
+        .withContext("Acting Parties")(_.addStringSet(exercise.actingParties))
+        .withContext("Interface Id")(_.addOptional(exercise.interfaceId, _.addIdentifier))
+        .withContext("Choice Id")(_.addString(exercise.choiceId))
+        .withContext("Chosen Value")(_.addWithValueBuilder(exercise.chosenValue, _.addTypedValue))
+        .withContext("Consuming")(_.addBool(exercise.consuming))
+        .withContext("Exercise Result")(
+          _.addOptional[Value](
+            exercise.exerciseResult,
+            { builder => value => builder.addWithValueBuilder(value, _.addTypedValue) },
+          )
+        )
+        .withContext("Choice Observers")(_.addStringSet(exercise.choiceObservers))
```

Comment:
What about the chosen value? What about choice authorizers?

Reply:
Chosen value is already [there](https://github.com/digital-asset/daml/pull/20172/files/1422f6c30058c9e5ee82a62b882fbf5c8322c6c7#diff-6cdb90c1758150a46910f9e1a187569615c4f998358046ad930f317bde3f50e5R369)

Reply:
I really should not review such PRs when I'm tired ;-)